Change the type of some pparam fields to align with how they are represented in Consensus

TimSheard opened the pull request:
Addresses issue  #3942, issue #3905, and issue #3906
Changes a few PParam fields to  Word32 and Word16 to align with how they are represents in Consensus.

Also changed 2 fields in the type BHeaderView, to align with the PParams changes

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

TimSheard committed:
Change the type of Conway PParams field cppCommitteeMaxTermLength (and related fields and Lenses) from EpochNo to EpochInterval

TimSheard committed:
Changed the ppMaxTxSize fields of PParams to Word32

TimSheard committed:
Changed the type of MaxBBSize (word32) and MaxBHSize (Word16).

lehins commented on CHANGELOG.md:
```diff
@@ -8,7 +8,8 @@ If you are looking for the Ledger Releasing and Versioning Process then you can
 ## 8.7
-
+-- Changed the underlying types of the 3 PParam fields MaxTxBodySize (to Word32) MaxBBodySize (to Word32) MaxBHeaderSize (to Word16)
+-- Changed the fields of BHeaderView to align with the changes in PParams	
```
This is a top level changelog that only lists information relevant to `cardano-node` users. So, neither of those changes are relevant for people who don't use ledger as a library. However, if they were relevant, they would have to be listed in a new section, since 8.7 has already been released and the next one will be 8.8
```suggestion

```

lehins commented on eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Generator/Update.hs:
```diff
@@ -261,9 +261,9 @@ genShelleyPParamsUpdate c@Constants {maxMinFeeA, maxMinFeeB, maxMajorPV} pp = do
-      & ppuMaxBHSizeL .~ maxBHSize
+      & ppuMaxBBSizeL .~ (fromIntegral <$> maxBBSize)
+      & ppuMaxTxSizeL .~ (fromIntegral <$> maxTxSize)
+      & ppuMaxBHSizeL .~ (fromIntegral <$> maxBHSize)
```
This guy will overflow. Since `200000` is much bigger that `65535`

I'd suggest creating a separate `genWord16` and `genWord32` instead of using `genNatural`

TimSheard replied:
So now I use explicit  choose(n,m) where n and m are word16, word32, or word64, that way I don't need to use genNatural.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/PParams.hs:
```diff
@@ -79,9 +79,9 @@ genPParams proof tx bb bh = do
-          , MaxBHSize bh
+          , MaxTxSize (fromIntegral tx) -- In the Model these are Natural
+          , MaxBBSize (fromIntegral bb) -- But in the PParams, they are
+          , MaxBHSize (fromIntegral bh) -- Word32, Word32, and Word16
```
Are there any guarantees they are not going to overflow when converted from a Natural?

It might be worth using something like that instead of `fromIntegral`:

```haskell
fromIntegralBounded :: (HasCallStack, Integral a, Bounded b, Num b) => a -> b
fromIntegralBounded x =
  | toInteger (minBound :: b) <= xi && toInteger (maxBound :: b) = fromIntegral x
  | otherwise = 
    error $ show x <> " is out of bounds: " <> show (minBound :: b, maxBound ::b)
  where
    xi = toInteger x
```

TimSheard replied:
So I did this
```
fromIntegralBounded :: forall a b.
  (HasCallStack, Integral a, Show a, Integral b, Bounded b, Show b) =>
  String -> a -> b
fromIntegralBounded name x
  | toInteger (minBound :: b) <= xi && xi <= toInteger (maxBound :: b) = fromIntegral x
  | otherwise = 
    error $ "While converting "++name++", "++show x <> " is out of bounds: " <> show (minBound :: b, maxBound ::b)
  where
    xi = toInteger x
```

This way the error tells what you were converting. Had to fix a few typos

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Vars.hs:
```diff
@@ -1099,9 +1099,15 @@ maxTxSize p =
+word32NaturalL = lens fromIntegral (\_ y -> fromIntegral (toInteger y))
+
+word16NaturalL :: Lens' Word16 Natural
+word16NaturalL = lens fromIntegral (\_ y -> fromIntegral (toInteger y))
```
Same here. These lenses could use some protection against overflow

TimSheard replied:
Done

lehins commented on libs/cardano-protocol-tpraos/src/Cardano/Protocol/TPraos/BHeader.hs:
```diff
@@ -354,16 +354,16 @@ prevHashToNonce = \case
-bHeaderSize :: forall c. BHeader c -> Int
-bHeaderSize = BS.length . bHeaderBytes
+bHeaderSize :: forall c. BHeader c -> Word16
+bHeaderSize = fromIntegral . BS.length . bHeaderBytes
```
This can easily overflow and that would be really bad! We need to switch it back to `Int`
```suggestion
bHeaderSize :: forall c. BHeader c -> Int
bHeaderSize = BS.length . bHeaderBytes
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/BHeaderView.hs:
```diff
@@ -22,9 +22,9 @@ data BHeaderView c = BHeaderView
+  , bhviewBSize :: Word32
   -- ^ The purported size (in bytes) of the block body.
-  , bhviewHSize :: Int
+  , bhviewHSize :: Word16
```
Switch it back to `Int`, since that is the field that used for reporting the actual size of the header. In this PR we want to change the type only for parameters that limit the size.
```suggestion
  , bhviewHSize :: Int
```

TimSheard committed:
Made changes to BHeaderView to align with changes to PParams
Adjust random generation of Tests, to meet reduced scope (Word16) of MaxBHSize.
Apply suggestions from code review Co-authored-by: Alexey Kuleshevich <alexey.kuleshevich@iohk.io>
Added function fromIntegralBounded.

lehins commented on CHANGELOG.md:
```diff
@@ -8,7 +8,6 @@ If you are looking for the Ledger Releasing and Versioning Process then you can
 [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd).
 
 ## 8.7
```
Bring back the newline
```suggestion
## 8.7

```

TimSheard replied:
done

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:
```diff
@@ -133,7 +133,7 @@ ppDRepVotingThresholdsL = ppLens . hkdDRepVotingThresholdsL @era @Identity
 ppCommitteeMinSizeL = ppLens . hkdCommitteeMinSizeL @era @Identity
 
-ppCommitteeMaxTermLengthL :: forall era. ConwayEraPParams era => Lens' (PParams era) EpochNo
+ppCommitteeMaxTermLengthL :: forall era. ConwayEraPParams era => Lens' (PParams era) EpochInterval
```
Change of type for `ppCommitteeMaxTermLengthL` and `ppuCommitteeMaxTermLengthL` warrants a changelog entry

TimSheard replied:
done

lehins replied:
Wrong changelog

TimSheard replied:
OK it now where it beongs

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Core/PParams.hs:
```diff
@@ -375,15 +376,15 @@ ppMinFeeBL :: forall era. EraPParams era => Lens' (PParams era) Coin
 
 -- | Maximal block body size
-ppMaxBBSizeL :: forall era. EraPParams era => Lens' (PParams era) Natural
+ppMaxBBSizeL :: forall era. EraPParams era => Lens' (PParams era) Word32
```
We need a changelog entry at least for the 6 lenses `ppMaxBBSizeL`, `ppMaxTxSizeL`, `ppMaxBHSizeL`, `ppuMaxBBSizeL`, `ppuMaxTxSizeL` and `ppuMaxBHSizeL`

Because they have a direct impact on the downstream users

TimSheard replied:
done

lehins replied:
Wring changelog

TimSheard replied:
Where they belong

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Chain.hs:
```diff
@@ -70,7 +70,7 @@ chainChecks maxpv ccd bhv = do
   unless (m <= maxpv) $ throwError (ObsoleteNodeCHAIN m maxpv)
   unless (fromIntegral (bhviewHSize bhv) <= ccMaxBHSize ccd) $
```
Here is another overflow, which we must guard against properly:
```suggestion
  unless (bhviewHSize bhv <= (fromIntegral :: Word16 -> Int) (ccMaxBHSize ccd)) $
```

Using `fromIntegral` without a type signature is really dangerous. We should start adding a type signature at every use site.

lehins reviewed (changes requested):
One more dangerous bug needs to be squashed and I think we are good to go.

lehins commented on CHANGELOG.md:
```diff
@@ -8,7 +8,8 @@ If you are looking for the Ledger Releasing and Versioning Process then you can
 ## 8.7
-
+	
+- Changed the type of lenses ppCommitteeMaxTermLengthL, ppuCommitteeMaxTermLengthL, ppMaxBBSizeL, ppMaxTxSizeL, ppMaxBHSizeL, ppuMaxBBSizeL, ppuMaxTxSizeL and ppuMaxBHSizeL
```
That is not the changelog that we need to change. Changelog for the package need to be adjusted.

```suggestion
```

TimSheard replied:
done

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Newpp.hs:
```diff
@@ -108,7 +108,7 @@ newPpTransition = do
           ?! UnexpectedDepositPot (Coin oblgCurr) (utxosDeposited utxoSt)
 
-      if (ppNew' ^. ppMaxTxSizeL + ppNew' ^. ppMaxBHSizeL) < (ppNew' ^. ppMaxBBSizeL)
+      if (ppNew' ^. ppMaxTxSizeL + fromIntegral (ppNew' ^. ppMaxBHSizeL) < ppNew' ^. ppMaxBBSizeL)
```
Although this is not really possible, since we would never submit such high values in Babbage (last era that runs `NEWPP` rule), there is a chance of overflow here. It is safer to convert to Integer. Here is another example of dangerous usage of `fromIntegral`
```suggestion
      if (toInteger (ppNew' ^. ppMaxTxSizeL) + toInteger (ppNew' ^. ppMaxBHSizeL) < toInteger (ppNew' ^. ppMaxBBSizeL))
```

TimSheard replied:
done

lehins commented on eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Generator/Update.hs:
```diff
@@ -242,9 +240,9 @@ genShelleyPParamsUpdate c@Constants {maxMinFeeA, maxMinFeeB, maxMajorPV} pp = do
-  maxBHSize <- genM $ genNatural low hi
+  maxBBSize <- genM $ choose (low, hi)
+  maxTxSize <- genM $ choose (low, hi)
+  maxBHSize <- genM $ choose (25000, 65535 :: Word16) -- Must stay in the range of Word16, but can't be too small
```
This way the intention is clearer
```suggestion
  -- Must stay in the range of Word16, but can't be too small
  maxBHSize <- genM $ choose (25000, maxBound :: Word16)
```

TimSheard replied:
done

lehins commented on eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Generator/Update.hs:
```diff
@@ -261,9 +259,9 @@ genShelleyPParamsUpdate c@Constants {maxMinFeeA, maxMinFeeB, maxMajorPV} pp = do
-      & ppuMaxBHSizeL .~ maxBHSize
+      & ppuMaxBBSizeL .~ (fromIntegral <$> maxBBSize)
+      & ppuMaxTxSizeL .~ (fromIntegral <$> maxTxSize)
+      & ppuMaxBHSizeL .~ (fromIntegral <$> maxBHSize)
```
Redundant conversion
```suggestion
      & ppuMaxBBSizeL .~ maxBBSize
      & ppuMaxTxSizeL .~ maxTxSize
      & ppuMaxBHSizeL .~ maxBHSize
```

TimSheard replied:
done

lehins commented on eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Generator/Update.hs:
```diff
@@ -138,14 +138,18 @@ genPParams c@Constants {maxMinFeeA, maxMinFeeB, minMajorPV, maxMajorPV} = do
-    rangeUpTo :: Natural -> Gen Natural
-    rangeUpTo upper = genNatural low upper
+        <$> choose (blockBodySize, blockBodySize `div` 2)
+        <*> choose (25000, 65535 :: Word16) -- Must stay in the range of Word16, but can't be too small
```
Ditto
```suggestion
        <*> choose (25000, maxBound :: Word16) -- Must stay in the range of Word16, but can't be too small
```

TimSheard replied:
done

lehins reviewed (approved):
Awesome. Thank you!
